### PR TITLE
Fix duplicate Travelers in Allowed Characters filter

### LINKED ISSUE #1080: "Use Equipped" count showing "-1" on Traveller
image

### LINKED ISSUE #1356: Duplicate Travelers in Allowed Characters filtering for Equipped Artifact Configuration
Adding a new traveler as a character will duplicate travelers multiple times in Allowed Characters for Equipped Artifact Configuration.

Steps to reproduce:

1. Add a traveler:

image
image
image
image

2. Go to any character, Optimize tab, and filter characters in "Equipped Artifact Configuration": 

image
image

3. Use any filter multiple times and find Travelers duplicated. In this case, I just filtered by "Lumine" in the name search:

image
image
image
image

image

### PR DESCRIPTION
**Pull request by priolette**
## Describe your changes

Fixed duplicate Travelers in Allowed Characters filter for Equipped Artifact Configuration. Now it shows only one Traveler, filters also account for only one, except for the element type fields.

## Issue or discord link

- Resolves #1356
- Resolves #1080

## Testing/validation

image
image
image

## Checklist before requesting a review (leave this PR as draft if any part of this list is not done.)

- [ ] I have commented my code, in hard-to understand areas.
- [ ] I have made corresponding changes to README or wiki.
- [ ] For front-end changes, I have updated the corresponding English translations.
- [x] Ran `yarn run mini-ci` locally to validate format + lint.
- [x] If there were format issues, I ran `nx format write` to resolve them automatically.
- [ ] If I have added a new library or app, I have updated the deployment scripts to ignore changes as needed

### CODE REVIEW

**Comment by nguyentvan7** [comment]
Note to self to check if this is fixed too: https://github.com/frzyc/genshin-optimizer/issues/1080

**Comment by priolette** [comment]
I made it worse actually
image
image

**Comment by priolette** [comment]
Same configuration after adjustments:

Screenshot 2024-01-29 at 12 24 41

It should also show all the other filters correctly now (for weapon types, elements and rarity)

File: apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx [hunk 1]
```diff
@@ -159,23 +170,46 @@ export default function AllowChar({
...
           const eleKey = sheet.elementKey
+          const weaponTypeKey = sheet.weaponTypeKey
+          const characterRarityKey = sheet.rarity
+
           ctMap.elementTotals[eleKey].total++
           if (charKeyMap[ck]) ctMap.elementTotals[eleKey].current++
 
-          const weaponTypeKey = sheet.weaponTypeKey
+          if (charKeyToLocCharKey(ck) === 'Traveler') {
+            console.log(charKeyMap)
```
Comment (nguyentvan7) [inline apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx:192]:
```suggestion
```

File: apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx [hunk 2]
```diff
@@ -159,23 +170,46 @@ export default function AllowChar({
...
+            if (!charKeyMap[ck] && locList.includes('Traveler')) {
```
Comment (nguyentvan7) [inline apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx:193]: Please add some comments to why we do this logic

**Review by nguyentvan7 — COMMENTED** [review]
Functionality looks good, just left some style comments

**Comment by priolette** [comment]
Sorry for the rushed PR, I would like to know if the logic could be improved somehow, I don't really like having two `if`s with the same conditionals, perhaps it would be possible to do this better?

File: apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx (hunk 2 above)
Comment (priolette) [reply apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx:197]: Added explanations, hopefully they are understandable

**Review by nguyentvan7 — COMMENTED** [review]
```typescript
          // Handle multiple Travelers
          if (charKeyToLocCharKey(ck) === 'Traveler') {
            // Add to each element if Traveler of that element is in the database
            //
            // This makes sure that if the Traveler is in the filter,
            // counts are added to each of their elements in the database
            if (!charKeyMap[ck] && locList.includes('Traveler')) {
              ctMap.elementTotals[eleKey].current++

              // Only count the traveler once for rarity and weapon type totals
              if (travelerProcessed) return
              ctMap.weaponTypeTotals[weaponTypeKey].current++
              ctMap.characterRarityTotals[characterRarityKey].current++
            }

            travelerProcessed = true
          }
```

**Review by nguyentvan7 — APPROVED** [review]

**Pull request merged**